File: packages/sync-cli/internal/tododb/tododb.go
// Package tododb encodes and decodes PalmOS To Do List's ToDoDB.pdb
// records and AppInfo block.
//
// Each todo record:
//
//	0   2  dueDate (UInt16) — packed Palm DateType, or 0xFFFF for none
//	         bits 0..4   = day (1..31)
//	         bits 5..8   = month (1..12)
//	         bits 9..15  = year offset from 1904
//	2   1  priority — bit 7 = completed, bits 0..6 = priority (1..5)
//	3   *  description — NUL-terminated UTF-8 / Palm Latin-1
//	..  *  notes        — NUL-terminated, optional
//
// Category index = record entry attributes low nibble (same as memos).
//
// AppInfo block is 282 bytes: 276 standard categories + 2 reserved +
// 2 padding + 2 dirtyAppInfo / sort order (we just preserve raw bytes
// past byte 276 to stay compatible with whatever the device wrote).
package tododb

import (
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
	"time"

	"github.com/palmvellum/palmvellum/packages/sync-cli/internal/memodb"
	"github.com/palmvellum/palmvellum/packages/sync-cli/internal/pdb"
)

const AppInfoLen = 282

// Todo is one to-do item.
type Todo struct {
	UniqueID    uint32
	Category    uint8
	DueDate     *time.Time // nil = no due date
	Priority    uint8      // 1..5; values outside the range clamp on encode
	Completed   bool
	Description string
	Notes       string
}

func decodeDate(d uint16) *time.Time {
	if d == 0xFFFF {
		return nil
	}
	year := int(d>>9) + 1904
	month := int((d >> 5) & 0xF)
	day := int(d & 0x1F)
	if month == 0 || day == 0 {
		return nil
	}
	t := time.Date(year, time.Month(month), day, 0, 0, 0, 0, time.UTC)
	return &t
}

func encodeDate(t *time.Time) uint16 {
	if t == nil {
		return 0xFFFF
	}
	y := t.Year() - 1904
	if y < 0 || y > 127 {
		return 0xFFFF
	}
	return uint16(y)<<9 | uint16(t.Month())<<5 | uint16(t.Day())
}

// DecodeTodo parses one record's bytes.
func DecodeTodo(data []byte) (Todo, error) {
	if len(data) < 4 {
		return Todo{}, fmt.Errorf("tododb: short record %d bytes", len(data))
	}
	t := Todo{}
	t.DueDate = decodeDate(binary.BigEndian.Uint16(data[0:2]))
	prioByte := data[2]
	t.Completed = prioByte&0x80 != 0
	t.Priority = prioByte & 0x7F

	descEnd := 3
	for descEnd < len(data) && data[descEnd] != 0 {
		descEnd++
	}
	t.Description = string(data[3:descEnd])

	if descEnd+1 < len(data) {
		notesStart := descEnd + 1
		notesEnd := notesStart
		for notesEnd < len(data) && data[notesEnd] != 0 {
			notesEnd++
		}
		t.Notes = string(data[notesStart:notesEnd])
	}
	return t, nil
}

// Encode serializes a Todo to record bytes.
func (t Todo) Encode() []byte {
	var buf bytes.Buffer
	_ = binary.Write(&buf, binary.BigEndian, encodeDate(t.DueDate))
	prio := t.Priority & 0x7F
	if prio == 0 {
		prio = 1
	} else if prio > 5 {
		prio = 5
	}
	if t.Completed {
		prio |= 0x80
	}
	buf.WriteByte(prio)
	buf.WriteString(t.Description)
	buf.WriteByte(0)
	if t.Notes != "" {
		buf.WriteString(t.Notes)
	}
	buf.WriteByte(0)
	return buf.Bytes()
}

// DecodeTodos walks the DB and decodes every record.
func DecodeTodos(db *pdb.DB) ([]Todo, error) {
	out := make([]Todo, 0, len(db.Records))
	for i, r := range db.Records {
		td, err := DecodeTodo(r.Data)
		if err != nil {
			return nil, fmt.Errorf("record %d: %w", i, err)
		}
		td.UniqueID = r.UniqueID
		td.Category = r.Attributes & 0x0F
		out = append(out, td)
	}
	return out, nil
}

// EncodeTodos turns Todo structs into pdb.Record entries.
func EncodeTodos(todos []Todo) []pdb.Record {
	out := make([]pdb.Record, 0, len(todos))
	for _, t := range todos {
		out = append(out, pdb.Record{
			UniqueID:   t.UniqueID,
			Attributes: t.Category & 0x0F,
			Data:       t.Encode(),
		})
	}
	return out
}

// ─── AppInfo ──────────────────────────────────────────────────

// AppInfo reuses memodb.AppInfo for categories — the layout is
// identical for the first 276 bytes. The remaining bytes (reserved
// + dirtyAppInfo + sort) are preserved verbatim when round-tripping
// an existing AppInfo, or zeroed for a fresh one.
type AppInfo struct {
	*memodb.AppInfo
	Tail []byte // bytes 276..281, preserved on encode
}

func ParseAppInfo(b []byte) (*AppInfo, error) {
	if len(b) < 276 {
		return nil, errors.New("tododb: appinfo too short")
	}
	cats, err := memodb.ParseAppInfo(b)
	if err != nil {
		return nil, err
	}
	ai := &AppInfo{AppInfo: cats}
	if len(b) > 276 {
		ai.Tail = append([]byte(nil), b[276:]...)
	}
	return ai, nil
}

func (ai *AppInfo) Encode() []byte {
	base := ai.AppInfo.Encode() // 280 bytes
	out := make([]byte, AppInfoLen)
	copy(out[0:280], base)
	if len(ai.Tail) > 0 {
		// overlay original tail past byte 276 (last 6 bytes max)
		copyLen := len(ai.Tail)
		if copyLen > AppInfoLen-276 {
			copyLen = AppInfoLen - 276
		}
		copy(out[276:276+copyLen], ai.Tail[:copyLen])
	}
	return out
}

func DefaultAppInfo() *AppInfo {
	return &AppInfo{
		AppInfo: memodb.DefaultAppInfo(),
		Tail:    make([]byte, AppInfoLen-276),
	}
}

// NewTodoDB constructs an empty ToDoDB skeleton.
func NewTodoDB(ai *AppInfo) *pdb.DB {
	return &pdb.DB{
		Name:    "ToDoDB",
		Type:    pdb.FourCC("DATA"),
		Creator: pdb.FourCC("todo"),
		AppInfo: ai.Encode(),
	}
}

File: packages/android-native/app/src/main/java/dev/tatliving/palmvellum/organizers/data/card/ToDoDb.kt
package dev.tatliving.palmvellum.organizers.data.card

/**
 * Encode/decode PalmOS To Do List's ToDoDB records and AppInfo block — a Kotlin
 * port of `packages/palm-engine/tododb/tododb.go`.
 *
 * Each todo record:
 *   0   2  dueDate (packed Palm DateType, 0xFFFF = none)
 *            bits 0..4 day(1..31), bits 5..8 month(1..12), bits 9..15 year-1904
 *   2   1  priority — bit7 = completed, bits 0..6 = priority (1..5)
 *   3   *  description (NUL-terminated)
 *   ..  *  notes (NUL-terminated, optional)
 * Category index = record-entry attributes low nibble (same as memos).
 */

const val TODO_APPINFO_LEN = 282

/** A due date as a plain calendar triple (no time zone — Palm dates are date-only). */
data class PalmDate(val year: Int, val month: Int, val day: Int) {
    /** ISO yyyy-MM-dd, matching the cloud metadata format. */
    fun iso(): String = "%04d-%02d-%02d".format(year, month, day)

    companion object {
        fun fromIso(s: String): PalmDate? {
            val p = s.split("-")
            if (p.size != 3) return null
            val y = p[0].toIntOrNull() ?: return null
            val m = p[1].toIntOrNull() ?: return null
            val d = p[2].toIntOrNull() ?: return null
            return PalmDate(y, m, d)
        }
    }
}

class Todo(
    var uniqueId: Int = 0,
    var category: Int = 0,
    var dueDate: PalmDate? = null,
    var priority: Int = 0, // 1..5; clamped on encode
    var completed: Boolean = false,
    var description: String = "",
    var notes: String = "",
)

/** ToDo AppInfo reuses the 276-byte category block, preserving bytes 276.. verbatim. */
class TodoAppInfo(
    val categories: MemoAppInfo,
    var tail: ByteArray = ByteArray(0),
) {
    fun encode(): ByteArray {
        val base = categories.encode() // 280 bytes
        val out = ByteArray(TODO_APPINFO_LEN)
        base.copyInto(out, 0, 0, minOf(base.size, 280))
        if (tail.isNotEmpty()) {
            val copyLen = minOf(tail.size, TODO_APPINFO_LEN - 276)
            tail.copyInto(out, 276, 0, copyLen)
        }
        return out
    }

    companion object {
        fun parse(b: ByteArray): TodoAppInfo? {
            if (b.size < 276) return null
            val cats = MemoAppInfo.parse(b) ?: return null
            val tail = if (b.size > 276) b.copyOfRange(276, b.size) else ByteArray(0)
            return TodoAppInfo(cats, tail)
        }

        fun default(): TodoAppInfo = TodoAppInfo(MemoAppInfo.default(), ByteArray(TODO_APPINFO_LEN - 276))
    }
}

object ToDoDb {

    private fun decodeDate(d: Int): PalmDate? {
        if (d == 0xFFFF) return null
        val year = (d ushr 9) + 1904
        val month = (d ushr 5) and 0xF
        val day = d and 0x1F
        if (month == 0 || day == 0) return null
        return PalmDate(year, month, day)
    }

    private fun encodeDate(t: PalmDate?): Int {
        if (t == null) return 0xFFFF
        val y = t.year - 1904
        if (y < 0 || y > 127) return 0xFFFF
        return (y shl 9) or (t.month shl 5) or t.day
    }

    fun decodeRecord(data: ByteArray): Todo {
        require(data.size >= 4) { "tododb: short record ${data.size} bytes" }
        val t = Todo()
        t.dueDate = decodeDate(((data[0].toInt() and 0xFF) shl 8) or (data[1].toInt() and 0xFF))
        val prioByte = data[2].toInt() and 0xFF
        t.completed = prioByte and 0x80 != 0
        t.priority = prioByte and 0x7F

        var descEnd = 3
        while (descEnd < data.size && data[descEnd].toInt() != 0) descEnd++
        t.description = PalmCharset.fromPalm(data.copyOfRange(3, descEnd))

        if (descEnd + 1 < data.size) {
            val notesStart = descEnd + 1
            var notesEnd = notesStart
            while (notesEnd < data.size && data[notesEnd].toInt() != 0) notesEnd++
            t.notes = PalmCharset.fromPalm(data.copyOfRange(notesStart, notesEnd))
        }
        return t
    }

    fun encodeRecord(t: Todo): ByteArray {
        val due = encodeDate(t.dueDate)
        var prio = t.priority and 0x7F
        if (prio == 0) prio = 1 else if (prio > 5) prio = 5
        if (t.completed) prio = prio or 0x80
        val desc = PalmCharset.toPalm(t.description)
        val notes = if (t.notes.isNotEmpty()) PalmCharset.toPalm(t.notes) else ByteArray(0)
        val out = ByteArray(2 + 1 + desc.size + 1 + notes.size + 1)
        out[0] = (due ushr 8).toByte()
        out[1] = due.toByte()
        out[2] = prio.toByte()
        desc.copyInto(out, 3, 0, desc.size)
        // NUL after description already 0; notes follow, then trailing NUL (already 0).
        notes.copyInto(out, 3 + desc.size + 1, 0, notes.size)
        return out
    }

    fun decode(db: PalmDb): List<Todo> = db.records.map { r ->
        decodeRecord(r.data).apply {
            uniqueId = r.uniqueId
            category = r.attributes and 0x0F
        }
    }

    fun encode(todos: List<Todo>): MutableList<PalmRecord> = todos.map { t ->
        PalmRecord(uniqueId = t.uniqueId, attributes = t.category and 0x0F, data = encodeRecord(t))
    }.toMutableList()

    fun newDb(ai: TodoAppInfo): PalmDb = PalmDb(
        name = "ToDoDB",
        type = "DATA".toByteArray(Charsets.US_ASCII),
        creator = "todo".toByteArray(Charsets.US_ASCII),
        appInfo = ai.encode(),
    )
}
